kame/kame/rip6query/rip6query.8
.\"	$KAME: rip6query.8,v 1.8 2006/10/25 05:45:59 jinmei Exp $
.\"
.\" Copyright (C) 1998 and 1999 WIDE Project.
.\" All rights reserved.
.\" 
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the project nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd October 7, 1999
.Dt RIP6QUERY 8
.Os KAME
.Sh NAME
.Nm rip6query
.Nd RIPng debugging tool
.\"
.Sh SYNOPSIS
.Nm
.Op Fl I Ar interface
.Op Fl P Ar source port
.Op Fl h Ar hop limit
.Op Fl p Ar prefix
.Op Fl s Ar source address
.Op Fl w Ar time
.Ar destination
.\"
.Sh DESCRIPTION
.Nm
requests remote RIPng daemon on
.Ar destination
to dump RIPng routing information.
It can also send a RIPng response message when
.Fl r
option is specified.
.Fl I
lets you specify outgoing
.Ar interface
for the query packet.
This was useful for a link-local address specified as
.Ar destination
to disambiguate the outgoing link, but now it is mostly obsoleted
because the link can be specified using the textual representation
for scoped addresses as defined in RFC4007.
When the
.Fl P
or
.fl s
option is specified, the source port or the source address for the
outgoing message packet is set to the specified value, respectively.
Similarly, the
.Fl h
option specifies the hop limit value of the IPv6 header for the
outgoing packet.
The
.Fl p
option specified a particular prefix for the request or response
message.  For a request message, it means a request for the route to
the destination specified by the prefix;
by default,
.Nm
requests the entire routing table of the remote router.
For a response message, this option must be given and specifies the
destination of the RTE.
.Fl w
specifies the time in seconds to wait for the initial
response from a gateway.
The default value is 5 seconds.
When the response message is generated,
this option is ignored and
.Nm
does not wait at all and 
.\"
.Sh SEE ALSO
.Xr route6d 8
.\"
.Sh HISTORY
The
.Nm
command first appeared in WIDE Hydrangea IPv6 protocol stack kit.


CHANGELOG
			CHANGELOG for KAME kit
$KAME: CHANGELOG,v 1.2836 2006/10/25 07:14:33 jinmei Exp $

<200610>
2006-10-25  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/route6d/route6d.c (riprecv): implemented more
	validation checks about incoming responses per RFC2080, and one
	additional heuristic check for safer operation.

2006-10-25  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/rip6query/rip6query.c (main): added some new options
	various options to customize the packet.

<200609>
2006-09-08  SUZUKI, Shinsuke <suz@alaxala.net>
	* kame/sys/netinet6/mldv2.c:
	  - stop updating MLDv1-compat-timer when an MLDv2 Query is received
	    in MLDv1-compat-mode (the 2006-09-06's commit is not enough)

2006-09-06  SUZUKI, Shinsuke <suz@alaxala.net>
	* kame/sys/netinet6/mldv2.c:
	  - stop updating MLDv1-compat-timer when an MLDv2 Query is received
	    in MLDv1-compat-mode (a bug due to a commit 2006-09-04)

	* freebsd5/sys/net/if_loop.c: 
	  let a loop-backed IPv6 packet go to lo0's BPF, while
	  software-emulated loopbacked ethernet frame still
	  goes to the physical I/F's BPF.

	  This closes the todo issue of the following item in 
	  CHANGELOG.2000.  

	  > Tue May 30 01:29:04 JST 2000  itojun@iijlab.net
	  ...
	  > XXX if_simloop() call from ip6_output() causes junk to be injected
	  > to bpf buffer.  we may need some fix separately from this commit.

	  (suggested by Max Laier)

2006-09-05  SUZUKI, Shinsuke <suz@alaxala.net>
	* kame/sys/netinet/in_msf.c, kame/sys/netinet6/in6_msf.c:
	does not search MSF by 'fmode' in case of SIOCGMSFILTER or 
	SIOCGIPMSFILTER.  (RFC3768 5.1.2 and 5.2.2)
	(suggested by Yu-Jung Lee)

2006-09-04  SUZUKI, Shinsuke <suz@alaxala.net>
	* kame/sys/netinet6/{in6_msf.c, mldv2.c}: fixed several bugs regarding
	  MLDv2 (suggested by Yu-Jung Lee)
	  - returns an error code when the MSF state change fails due to a
	    lack of memory
	  - fixed a bug that an MLDv2 State-Change Report is not always
	    advertised robustness-variables times.
	  - MLDv1-compat-timer properly checks the current compat-timer
	    status

<200608>
2006-08-29  SUZUKI, Shinsuke <suz@alaxala.net>
	* kame/sys/netinet/icmp6.h: use ifdef for IGMPv3/MLDv2-related state
	macros to prevent a macro duplication in userland applications.
	(suggested by Hitoshi Asaeda)

<200607>
2006-07-11  SUZUKI, Shinsuke <suz@alaxala.net>
	* kame/sys/netinet/in_msf.c: fixed a bug that MSF advanced-API
	is rejected in case of IPv4.
	* kame/sys/netinet/igmp.c: fixed a kernel crash in freebsd5 when
	MSF advanced-API is used in a kernel with IGMPV3 option on.

<200606>
2006-06-09  Tsuyoshi MOMOSE  <momose@az.jp.nec.com>
	* kame/kame/shisad: RFC4285 on home agnets is implemented.
	
<200604>
2006-04-23  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/sys/netinet6/nd6.c (nd6_is_addr_neighbor): disabled the
	on-link assumption from the determination logic.

2006-04-09  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/sys/netinet6/ip6_mroute.c (ip6_mrouter_done): used
	if_allmulti(0) for diabling the multicast promiscuous mode for
	FreeBSD.  Without this change, an IPv6 multicast router would
	leave interfaces in the promiscuous mode even after the routing
	daemon terminates.

2006-04-02  SUZUKI, Shinsuke <suz@alaxala.net>
	* kame/sys/netinet6/mldv2.c: fixed a bug that an MLDv2 IS-IN Report
	is not advertised for an MLDv2 general Query

<200603>
2006-03-28  SUZUKI, Shinsuke <suz@alaxala.net>
	* kame/sys/netinet6/ip6_output.c: fixed a kernel crash by an ASM
	join from a userland application (KAME PR sys/932)

2006-03-27  SUZUKI, Shinsuke <suz@alaxala.net>
	* kame/sys/netinet6/mldv2.c: fixed several bugs regarding MLDv2-timer

2006-03-22  SUZUKI, Shinsuke <suz@alaxala.net>
	* kame/sys/netinet6/nd6.c: fixed a kernel crash due to a double-free
	introduced in the below fix

2006-03-17  SUZUKI, Shinsuke <suz@alaxala.net>
	* kame/sys/netinet6/nd6.c: fixed a memory leak for an unresolved
	packet in ND queue

2006-03-15  SUZUKI, Shinsuke <suz@alaxala.net>
	* kame/sys/netinet6/nd6_rtr.c: implements section 2.2. of 
	RFC4191, regarding the reserved preference value (10).
	
Fri Mar  3 20:20:37 JST 2006	keiichi@iijlab.net
	* netbsd/sys/netinet6/in6_pcb.c: send an IPV6_PATHMTU ancillary
	data when sending a packet whose size is larger than the path
	mtu to the destination.

2005-03-03  SUZUKI, Shinsuke <suz@alaxala.net>
	* removed the dhcpv6-related programs, since they are 
	  now maintained/developed/released in wide-dhcpv6 project in sourceforge.net.

2006-03-03  SUZUKI, Shinsuke <suz@alaxala.net>
	* kame/sys/netinet6/mldv2.c (mld_state_change_report)
	an update to the fix in Feb 11, 2006

<200602>
2006-02-24  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/sys/netinet6/icmp6.c (icmp6_redirect_input): fixed a bug
	that a Redirect couldn't be processed correctly if the original
	packet contained an extension header.
	Pointed out by: Qing Li

2006-02-11  SUZUKI, Shinsuke <suz@alaxala.net>
	* kame/sys/netinet6/mldv2.c (mld_state_change_report)
	fixed a bug that no MLDv2 report is advertised when host state changes 
	from IN/EX to EX/IN, respectively.
	Reported by: Yu-Jung Lee

2006-02-01  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* freebsd5/sys/kern/uipc_mbuf.c (m_pullup): fixed memory leak when
	the original mbuf had an mtag and the pull-up operation required
	allocating a new mbuf.
	Reported by: Garfield Hsieh

<200601>
2006-01-15  SUZUKI, Shinsuke <suz@alaxala.net>
	* kame/sys/netinet6/{icmp6.c, in6.[ch]}: not reply an ICMPv6 Error
	message for a packet from an obvious anycast address.

2006-01-09  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* freebsd5/sys/netinet6/in6_pcb.c (in6_{set,peer}sockaddr):
	converted embedded scope zone IDs into the sin6_scope_id field
	appropriately.
	Reported by: Susan.Lan

<200511>
2005-11-29  Tsuyoshi MOMOSE  <momose@az.jp.nec.com>
	* kame/kame/shisad, kame/sys/netinet6/{mip6.c,nd6_nbr.c}:
	Proxy DAD on a home agent is supported.

<200510>

kame/kame/rip6query/rip6query.c
/*	$KAME: rip6query.c,v 1.20 2006/10/25 05:45:59 jinmei Exp $	*/

/*
 * Copyright (C) 1995, 1996, 1997, and 1998 WIDE Project.
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the project nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <stdio.h>

#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <signal.h>
#include <errno.h>
#include <err.h>

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/queue.h>

#include <net/if.h>
#ifdef __FreeBSD__
#include <net/if_var.h>
#endif
#include <netinet/in.h>
#include <netinet/in_var.h>
#include <arpa/inet.h>
#include <netdb.h>

#include "route6d.h"

#define	DEFAULT_WAIT	5

int	s;
int	query_wait = DEFAULT_WAIT;
struct sockaddr_in6 sin6;
struct rip6	*ripbuf;

#define	RIPSIZE(n)	(sizeof(struct rip6) + (n-1) * sizeof(struct netinfo6))

int main __P((int, char **));
static void usage __P((void));
static void sigalrm_handler __P((int));
static const char *sa_n2a __P((struct sockaddr *));
static const char *inet6_n2a __P((struct in6_addr *));

int
main(argc, argv)
	int argc;
	char **argv;
{
	struct netinfo6 *np;
	struct sockaddr_in6 fsock;
	int i, n, len;
	socklen_t flen;
	int c;
	int ifidx = -1;
	int hlim = -1;
	u_char rip6_cmd = RIP6_REQUEST;

	int error;
	char pbuf[NI_MAXSERV];
	char *prefixstr = NULL;
	char *srcaddr = NULL;
	char *srcport = NULL;
	struct addrinfo hints, *res;

	while ((c = getopt(argc, argv, "I:w:p:P:h:rs:")) != -1) {
		switch (c) {
		case 'I':
			ifidx = if_nametoindex(optarg);
			if (ifidx == 0) {
				errx(1, "invalid interface %s", optarg);
				/*NOTREACHED*/
			}
			break;
		case 'P':
			srcport = optarg;
			break;
		case 'h':
			hlim = atoi(optarg);
			if (hlim <= 0 || hlim > 255)
				errx(1, "invalid hop limit: %s", optarg);
			break;
		case 'p':
			prefixstr = optarg;
			break;
		case 'r':
			rip6_cmd = RIP6_RESPONSE;
			break;
		case 's':
			srcaddr = optarg;
			break;
		case 'w':
			query_wait = atoi(optarg);
			break;
		default:
			usage();
			exit(1);
			/*NOTREACHED*/
		}
	}
	argv += optind;
	argc -= optind;

	if (argc != 1) {
		usage();
		exit(1);
	}

	if (rip6_cmd == RIP6_RESPONSE && prefixstr == NULL)
		errx(1, "-p must be specified for a response");

	if ((s = socket(AF_INET6, SOCK_DGRAM, 0)) < 0) {
		err(1, "socket");
		/*NOTREACHED*/
	}

	if (hlim > 0) {
		 if (setsockopt(s, IPPROTO_IPV6, IPV6_UNICAST_HOPS,
		     &hlim, sizeof(hlim)) == -1) {
			 err(1, "setsockopt(IPV6_UNICAST_HOPS)");
		 }
		 if (setsockopt(s, IPPROTO_IPV6, IPV6_MULTICAST_HOPS,
		     &hlim, sizeof(hlim)) == -1) {
			 err(1, "setsockopt(IPV6_MULTICAST_HOPS)");
		 }
	}

	if (srcaddr != NULL || srcport != NULL) {
		memset(&hints, 0, sizeof(hints));
		hints.ai_family = AF_INET6;
		hints.ai_flags = AI_PASSIVE;
		hints.ai_socktype = SOCK_DGRAM;
		hints.ai_protocol = IPPROTO_UDP;
		error = getaddrinfo(srcaddr, srcport, &hints, &res);
		if (error) {
			errx(1, "getaddrinfo failed for [%s]:%s: %s",
			    srcaddr ? srcaddr : "(unspec)",
			    srcport ? srcport : "(unspec)",
			    gai_strerror(error));
		}
		if (bind(s, res->ai_addr, res->ai_addrlen) < 0)
			err(1, "bind");
		freeaddrinfo(res);
	}

	/* getaddrinfo is preferred for addr%scope syntax */
	snprintf(pbuf, sizeof(pbuf), "%d", RIP6_PORT);
	memset(&hints, 0, sizeof(hints));
	hints.ai_family = AF_INET6;
	hints.ai_socktype = SOCK_DGRAM;
	hints.ai_protocol = IPPROTO_UDP;
	error = getaddrinfo(argv[0], pbuf, &hints, &res);
	if (error) {
		errx(1, "%s: %s", argv[0], gai_strerror(error));
		/*NOTREACHED*/
	}
	if (res->ai_next) {
		errx(1, "%s: %s", argv[0], "resolved to multiple addrs");
		/*NOTREACHED*/
	}
	if (sizeof(sin6) != res->ai_addrlen) {
		errx(1, "%s: %s", argv[0], "invalid addrlen");
		/*NOTREACHED*/
	}
	memcpy(&sin6, res->ai_addr, res->ai_addrlen);
	if (ifidx >= 0)
		sin6.sin6_scope_id = ifidx;

	if ((ripbuf = (struct rip6 *)malloc(BUFSIZ)) == NULL) {
		err(1, "malloc");
		/*NOTREACHED*/
	}
	/*ripbuf->rip6_cmd = RIP6_REQUEST;*/
	ripbuf->rip6_cmd = rip6_cmd;
	ripbuf->rip6_vers = RIP6_VERSION;
	ripbuf->rip6_res1[0] = 0;
	ripbuf->rip6_res1[1] = 0;
	np = ripbuf->rip6_nets;
	np->rip6_tag = 0;

	if (prefixstr != NULL) {
		char *cp;
		int plen;

		cp = strchr(prefixstr, '/');
		if (cp == NULL)
			errx(1, "invalid prefix format: prefixstr: %s",
			    prefixstr);
		*cp = '\0';
		if (inet_pton(AF_INET6, prefixstr, &np->rip6_dest) != 1)
			errx(1, "inet_pton failed for %s", prefixstr);
		plen = atoi(cp + 1);
		if (plen < 0 || plen > 128)
			errx(1, "invalid prefix length: %s\n", cp + 1);
		np->rip6_plen = (u_char)plen;
		np->rip6_metric = 1; /* dummy metric */
	} else {
		bzero(&np->rip6_dest, sizeof(struct in6_addr));
		np->rip6_plen = 0;
		np->rip6_metric = HOPCNT_INFINITY6;
	}
	if (sendto(s, ripbuf, RIPSIZE(1), 0, (struct sockaddr *)&sin6,
	    sizeof(struct sockaddr_in6)) < 0) {
		err(1, "send");
		/*NOTREACHED*/
	}
	if (rip6_cmd == RIP6_RESPONSE)
		exit(0);
	signal(SIGALRM, sigalrm_handler);
	for (;;) {
		flen = sizeof(fsock);
		alarm(query_wait);
		if ((len = recvfrom(s, ripbuf, BUFSIZ, 0,
		    (struct sockaddr *)&fsock, &flen)) < 0) {
			err(1, "recvfrom");
			/*NOTREACHED*/
		}
		alarm(0);
		printf("Response from %s len %d\n",
		    sa_n2a((struct sockaddr *)&fsock), len);
		n = (len - sizeof(struct rip6) + sizeof(struct netinfo6)) /
		    sizeof(struct netinfo6);
		np = ripbuf->rip6_nets;
		for (i = 0; i < n; i++, np++) {
			printf("\t%s/%d [%d]", inet6_n2a(&np->rip6_dest),
			    np->rip6_plen, np->rip6_metric);
			if (np->rip6_tag)
				printf(" tag=0x%x", ntohs(np->rip6_tag));
			printf("\n");
		}
	}

	exit(0);
}

static void
usage()
{
	fprintf(stderr,
	    "Usage: rip6query [-r] [-I iface] [-w wait] [-p prefix]\n"
	    "                 [-h hop_limit] [-s source_address] "
	    "[-P source_port] address\n");
}

/* getnameinfo() is preferred as we may be able to show ifindex as ifname */
static const char *
sa_n2a(sa)
	struct sockaddr *sa;
{
	static char buf[NI_MAXHOST];

	if (getnameinfo(sa, sa->sa_len, buf, sizeof(buf),
	    NULL, 0, NI_NUMERICHOST) != 0) {
		snprintf(buf, sizeof(buf), "%s", "(invalid)");
	}
	return buf;
}

static const char *
inet6_n2a(addr)
	struct in6_addr *addr;
{
	static char buf[NI_MAXHOST];

	return inet_ntop(AF_INET6, addr, buf, sizeof(buf));
}

static void
sigalrm_handler(sig)
	int sig;
{

	_exit(0);
}
